BUG: numpy.einsum indexing arrays now accept numpy int type

rlintott opened the pull request:
Fixes #15961.

All current tests pass. I didn't write any new tests for this though, but perhaps I should do that?

rlintott committed:
BUG: numpy.einsum indexing arrays now accept numpy int type

See #15961.

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -2459,46 +2458,48 @@ einsum_list_to_subscripts(PyObject *obj, char *subscripts, int subsize)
-            long s = PyInt_AsLong(item);
-            npy_bool bad_input = 0;
+        else {
+            long s = PyLong_AsLong(item);
```
What error message does this give if you pass a float? If it doesn't error at all, you might need to change to the following:
```suggestion
            Py_ssize s = PyNumber_Index(item);
```

rlintott replied:
You're right it didn't give an error on floats. I think I have fixed that now. Hopefully 

`item = PyNumber_Index(item);` 

is OK? Since it just modifies the object. I'm very new to the C Python API.

eric-wieser replied:
Oh, I guess I forgot the return type. I think we might have a helper function that does what we need here, perhaps `PyArray_IntAsIntp` or something similar.

seberg replied:
Yes, which is much the same as `PyNumber_AsSsize_t ` but rejects bools.

rlintott replied:
I have edited the function to use `PyArray_PyIntAsIntp` instead, which is what I believe you were referencing. It did used to accept bools before, but now it doesn't.

seberg replied:
Yeah, I am not quite sure. Doesn't feel quite right to switch over silently, but on the other hand it seems not like a much used corner and then a corner case of that.

Can you add those new tests in either case? They are necessary whether or not we let bools pass.

rlintott committed:
BUG: numpy.einsum indexing arrays now accept numpy int type

See #15961

rlintott committed:
Using PyArray_PyIntAsIntp helper function instead

rlintott committed:
TST: add tests for einsum numpy int and bool list subscripts

Added tests to check that einsum accepts numpy int64 types and
rejects bool. Rejecting bools is new behaviour in subscript lists.
I changed ValueError to TypeError on line 2496 in multiarraymodule.c
as it is more appropriate. I also modified einsumfunc.py to have the
same behaviour as in the C file when checking subscript list.
(Reject bools but accept anything else from operator.index())

rlintott commented:
Please let me know what you think of this new commit! I had to change `einsumfunc.py` as I did not realise there were two implementations until I started running tests. It does reject bools now. I also changed a valueerror to a typeerror as it seemed more appropriate (and it was causing my tests to fail)

eric-wieser commented on numpy/core/tests/test_einsum.py:
```diff
@@ -93,6 +93,10 @@ def test_einsum_errors(self):
                 b = np.ones((3, 4, 5))
                 np.einsum('aabcb,abc', a, b)
+            
+            #do not accept bools in subscript list
```
```suggestion
            # do not accept bools in subscript list
```
PEP8

eric-wieser commented on numpy/core/tests/test_einsum.py:
```diff
@@ -274,6 +278,13 @@ def check_einsum_sums(self, dtype, do_opt=False):
             assert_equal(np.einsum(a, [0, 0], optimize=do_opt),
                          np.trace(a).astype(dtype))
 
+            #ticket #15961: should accept numpy int64 type in subscript list
```
```suggestion
            # gh-15961: should accept numpy int64 type in subscript list
```
"Ticket" refers to a previous issue tracker

seberg commented on numpy/core/einsumfunc.py:
```diff
@@ -576,7 +577,7 @@ def _parse_einsum_input(operands):
                 if s is Ellipsis:
                     subscripts += "..."
-                elif isinstance(s, int):
+                elif isinstance(operator.index(s), int) and not isinstance(s, bool):
```
yes, exposing the c converter for these privately seems fine to me. I am also OK with just accepting bools for now here, doing the same in C,,, Its auper important for indexing-like things, oherwise I still like it, but it seams not too important,

eric-wieser replied:
I'd also be ok with being inconsistent about accepting bools for now between C and python, and then fix that type of issue numpy-wide later.

rlintott committed:
Applying requested changes

eric-wieser commented on numpy/core/einsumfunc.py:
```diff
@@ -590,11 +592,13 @@ def _parse_einsum_input(operands):
+                        s = operator.index(s)
+                    except TypeError as e:
+                        raise TypeError("For this input type lists must contain "
+                                        "either int or Ellipsis")
```
```suggestion
                                        "either int or Ellipsis") from e
```

eric-wieser commented on numpy/core/einsumfunc.py:
```diff
@@ -577,11 +577,13 @@ def _parse_einsum_input(operands):
+                        s = operator.index(s)
+                    except TypeError as e:
+                        raise TypeError("For this input type lists must contain "
+                                        "either int or Ellipsis")
```
```suggestion
                                        "either int or Ellipsis") from e
```

rlintott committed:
improve exception handling

anirudh2290 commented on numpy/core/tests/test_einsum.py:
```diff
@@ -274,6 +274,13 @@ def check_einsum_sums(self, dtype, do_opt=False):
+            np_array = np.asarray([0, 0])
+            assert_equal(np.einsum(a, np_array, optimize=do_opt),
+                         np.trace(a).astype(dtype))
+            assert_equal(np.einsum(a, list(np_array), optimize=do_opt),
```
Are we testing for other opt_types for np.int64 and optimize values which are not False. This may be useful to trigger einsum_path code path and ensure _parse_einsum_input is correct.

rlintott replied:
Sorry, what do you mean by other opt_types? I am very new to working on numpy

But it does test for both optimize=False and optimize=True, and goes through both the C and Python implementation if that is what you mean

anirudh2290 replied:
apologies for the delay in reply, what i meant is adding a `self.check_einsum_sums('i8', True)` for `test_einsum_sums_int64`. by other opt_types i mean other options for optimize (False, True, greedy, optimal) (likely False, True, check may suffice).

rlintott replied:
Ahh, I think I see. It seems like `_parse_einsum_input` is called in all cases if optimization is turned on, so I think the test coverage is complete. I have put the subscript list test in `self.check_einsum_sums` itself so that it is tested with all dtypes and opt_types that were being tested before. Hopefully this is OK

anirudh2290 replied:
Yes, on a second look I think this test handles other values for optimize and I think it should be fine.

anirudh2290 reviewed (commented):
Thanks for working on this, overall looks good !

charris commented:
Could you add a release note for this under `Improvements`.

rlintott committed:
release note added

seberg commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -2459,46 +2458,49 @@ einsum_list_to_subscripts(PyObject *obj, char *subscripts, int subsize)
+        else {
+            npy_intp s = PyArray_PyIntAsIntp(item);
+            /* Subscript */
+            if (!PyErr_Occurred()) {
```
```suggestion
            if (!error_converting(s)) {
```
A bit of a micro optimization, but its good to use generally and the more places use it, the more likely we don't forget it... (These functions return -1 on failure, so checking that first skips the `PyErr_Occured()` check in most cases).

Anyway, only glanced over but Anirudh and Eric had a look, so planning on merging it later. (I will apply this change myself).

seberg replied:
It might be a bit nicer to move the error into this if rather than the opposite way and then keep the rest unindented (as it was). I also feel its a bit more standard pattern handle the error immediately.

seberg commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -2459,46 +2458,49 @@ einsum_list_to_subscripts(PyObject *obj, char *subscripts, int subsize)
-            else {
-                bad_input = 1;
-            }
+                if ( s < 0 ) {
```
```suggestion
                if (s < 0) {
```
Ah, was there before... but I think I will commit it anyway...

seberg reviewed (commented):
Just some nits, I may just commit those suggestions and merge in a day or so. If you have some time, happy if you finish it off though (and please see if moving the error creation looks better)!

rlintott committed:
applying sebergs suggestions

seberg commented:
Thanks Ryan, lets put it in before I wonder if it was nicer before ;).
